GS: Add a Vulkan renderer

stenzek opened the pull request:
### Description of Changes

What the title says :)

Been working on this for a while, but it would be nice to get some testing from a wider range of devices. Originally it was tied to the Qt PRs, but I've moved the commits around so it's qt-base-2 > vulkan -> full qt interface.

The main advantages to Vulkan over Direct3D/GL:
 - Speed. Vulkan can do handle high blending loads where even NVIDIA's GL driver struggles.
 - Future-proofiing.
 - In theory, more efficient than D3D/GL.

Scary line count is due to pulling in VulkanMemoryAllocator and the vulkan headers.

A note for anyone trying to start silly licensing drama: Yes, some of the wrapper code is originally from Dolphin and then Duckstation. In the former case, it's not the most recent code, it's based on the original code I myself contributed back in 2016, with large modifications, and I never signed a CLA. So I still have copyright over it and can re-license as I see fit (to LGPL here).

### Suggested Testing Steps

Try blending games :) I've only been working on AMD for a while, hopefully the performance is still good on NVIDIA. No idea about Intel.

Currently, AMD is much slower than NVIDIA for blending. See refraction's post here: https://community.amd.com/t5/opengl-vulkan/vulkan-poor-performance-due-to-barrier-region-bit-being-ignored/td-p/501962

.. but it's still twice as fast as OpenGL on AMD :)

stenzek commented:
Will sort out the linux issues tomorrow when I have time to actually boot into linux, should be fine for testing after then.

RedDevilus commented:
Windows issue when opening Graphics settings while running a game:
```
SetPixelFormat() failed: 0xC0070006
Failed to switch to surfaceless
```

vitorhnn commented:
Built and did some quick (~5 min) testing on Linux w/ RADV, seems to mostly work, but required a small patch:
```diff
diff --git a/pcsx2/GS/Renderers/Vulkan/GSDeviceVK.cpp b/pcsx2/GS/Renderers/Vulkan/GSDeviceVK.cpp
index 14d56cb1f..36ced2592 100644
--- a/pcsx2/GS/Renderers/Vulkan/GSDeviceVK.cpp
+++ b/pcsx2/GS/Renderers/Vulkan/GSDeviceVK.cpp
@@ -30,6 +30,7 @@
 #include "HostDisplay.h"
 #include <sstream>
 #include <climits>
+#include <cfloat>
 
 #ifdef ENABLE_OGL_DEBUG
 static u32 s_debug_scope_depth = 0;
diff --git a/pcsx2/GS/Renderers/Vulkan/GSTextureVK.h b/pcsx2/GS/Renderers/Vulkan/GSTextureVK.h
index 8c9d19b24..fd4b5e169 100644
--- a/pcsx2/GS/Renderers/Vulkan/GSTextureVK.h
+++ b/pcsx2/GS/Renderers/Vulkan/GSTextureVK.h
@@ -24,14 +24,8 @@ class GSTextureVK final : public GSTexture
 public:
 	union ClearValue
 	{
-		struct
-		{
-			float depth;
-		};
-		struct
-		{
-			GSVector4 color;
-		};
+		float depth;
+		GSVector4 color;
 	};
 
 public:
```

refractionpcsx2 commented:
Weird things I had happen in a short time testing.

1. If you escape the emulation to close the GS window, then go in to the settings, I think I changed the OSD settings, then Okayed it to come out, it opens a dead GS window, even though one shouldn't be open at this point.  Although going to System-Resume does bring it back to life.

2. Trying to go in to the GS settings while the emulation is running

Vulkan: prints "Writing 506645 bytes to 'G:\ps2\gitrepo\vulkan\cache\vulkan_pipelines.bin'" then crashes to desktop
OpenGL: prints 
"SetPixelFormat() failed: 0xC0070006
Failed to switch to surfaceless"
but does successfully go in to the options.

TransparentBlue commented:
Anything I've tested (MGSS2 / MGS3 / Rayman 2R / GT4 ) that either requires high / full blending for some effects and / or has massive performance drops from enabling blending at all ( #2355 ) gets anywhere from 1.25x to 4x framerate, MGS2/3 cutscenes being on the lower end of that Rayman 2R being on the upper end of that, no graphical difference that I could see either.

Performance regression on Armored Core 3 not sure why.
![AC3 ogl](https://user-images.githubusercontent.com/16150597/147617979-7c23164d-e33b-42a7-bdd4-a29834c05869.png)
![AC3 vk](https://user-images.githubusercontent.com/16150597/147617983-c65c3c2d-1305-436a-ab84-99ed6634e2a4.png)
GS dump for that scene
[AC3 intro.gs.xz.zip](https://github.com/PCSX2/pcsx2/files/7786907/AC3.intro.gs.xz.zip)
Thought it might have to do with depth emulation but disabling it makes the performance difference even bigger
![AC3 ogl depth off](https://user-images.githubusercontent.com/16150597/147618012-5bb25218-cd9d-48b7-b75d-f3cfa76602ca.png)
![AC3 vk depth off](https://user-images.githubusercontent.com/16150597/147618013-a25c1a29-e16b-4186-9cc8-e66053a31806.png)

arcum42 commented:
Testing on Linux with the patches above applied:

When quitting while the emulator is running, it segfaults on vkUnmapMemory(g_vulkan_context->GetDevice(), m_memory); in Vulkan::StreamBuffer::Destroy. (Line 150 of StreamBuffer.cpp).

MrCK1 commented:
- Tested on Windows. I am able to reproduce the crash opening graphics settings.
- Mild performance improvement across games bottlenecked by blending. 

Did notice a bug when HW upscaling in Hitman Contracts - the pixels of the menu font can be seen "swimming" back and forth. Does not occur with OpenGL SW/HW. 

https://user-images.githubusercontent.com/16314399/147621582-5ac2e89c-c513-4c4b-b401-f725a123864d.mp4



Maybe the same bug as Mission Impossible here -
![pcsx2-avx2_2021_12_28_21_18_40_856](https://user-images.githubusercontent.com/16314399/147621343-d715f0c2-c149-4200-b3cf-ba8d676abad8.png)

GS Dump (rename to .xz)
[mi_vulkan.zip](https://github.com/PCSX2/pcsx2/files/7787075/mi_vulkan.zip)

tellowkrinkle commented:
Closing Vulkan GSdump playback in the GS debuggercrashes PCSX2

stenzek commented:
> Anything I've tested (MGS2 / MGS3 / Rayman 2R / GT4 ) that either requires high / full blending for some effects and / or has massive performance drops from enabling blending at all ( #2355 ) gets anywhere from 1.25x to 4x framerate, MGS2/3 cutscenes being on the lower end of that Rayman 2R being on the upper end of that, no graphical difference that I could see either.

You didn't post what your GPU is.

TransparentBlue commented:
>You didn't post what your GPU is.
Sorry about that/
Nvidia GTX 1080ti w/ 472.12 drivers

Dreadmoth commented:
> Shadows are broken in Ratchet & Clank 

Strange, they work fine for me on OpenGL and Vulkan with either method:
• Minimum blending accuracy and auto flush on at native resolution
• Basic or higher blending at native or any higher resolution

Ratchet & Clank PAL SCES-50916
i7-4770, GTX1070 Ti (driver 497.29)
Windows 10 21H2

arcum42 commented:
After the last force push, linux compiles without issues, and the segfault I listed on quit no longer happens.

refractionpcsx2 commented:
> > Shadows are broken in Ratchet & Clank
> 
> Strange, they work fine for me on OpenGL and Vulkan with either method: • Minimum blending accuracy and auto flush on at native resolution • Basic or higher blending at native or any higher resolution
> 
> Ratchet & Clank PAL SCES-50916 i7-4770, GTX1070 Ti (driver 497.29) Windows 10 21H2

yeah it seems barriers just aren't working for me at all on my 2080Ti.. gonna try updating my drivers (I'm on like 494)

refractionpcsx2 commented:
Driver update fixed Ratchet

MrCK1 commented:
Disgregard my post, my issues were fixed by updating Nvidia drivers as well

Dreadmoth commented:
Snowblind Engine seems to be unhappy - graphics are fine, but performance can be halved in some scenes:

[Champions: Return to Arms OpenGL - 95FPS](https://user-images.githubusercontent.com/8492684/147626934-e2dbe26f-9c60-4982-bd9b-6b46860497c4.png)

[Champions: Return to Arms Vulkan - 49FPS](https://user-images.githubusercontent.com/8492684/147626936-1ba5f0ab-3a5b-4a17-b945-bec8b6aac579.png)

(similar results in Baldur's Gate: Dark Alliance and Fallout: Brotherhood of Steel)

stenzek commented:
> Snowblind Engine seems to be unhappy - graphics are fine, but performance can be halved in some scenes:
> 
> [Champions: Return to Arms OpenGL - 95FPS](https://user-images.githubusercontent.com/8492684/147626934-e2dbe26f-9c60-4982-bd9b-6b46860497c4.png)
> 
> [Champions: Return to Arms Vulkan - 49FPS](https://user-images.githubusercontent.com/8492684/147626936-1ba5f0ab-3a5b-4a17-b945-bec8b6aac579.png)
> 
> (similar results in Baldur's Gate: Dark Alliance and Fallout: Brotherhood of Steel)
> 
> i7-4770, GTX1070 Ti (driver 497.29) Windows 10 21H2

Possibly the texture uploads. I have another feature (not currently in this PR) which makes it much more efficient for vulkan, but still needs some work with handling mipmaps.

Risae commented:
Found 2 issues, those do not seem to exist in the OpenGL or DX11 Renderer.

OS: Windows 10 Pro 21H2, Build 19044.1415
CPU: AMD Ryzen 7 2700X
GPU: Nvidia RTX 2070, Driver version 497.29

Game: Growlanser VI: Precarious World SLPM-66716
PCSX2 version: PCSX2-64bit-AVX2-pr[5224]-sha[442e1f7]-title[[WIP] GS Add a Vulkan renderer]
PCSX2 Settings:
1. "Clear All Settings" and restarted PCSX2
2. Config > Graphics Settings: Renderer changed to "Vulkan"
3. Config > Audio Settings: Changed Audio Module to "Cubeb"
4. Config > GamePad Settings: Imported PAD File through "Load Bindings"
5. System: Disable "Fast Boot"
6. System > Game Settings: Enable "Enable Cheats" (in order to go to the debug room of the game)

**Issue 1 - Flickering of the screen background textures:**
Video: https://www.youtube.com/watch?v=bdWb7Rjxexw
GS Dump: [gs_20211229062410.gs.xz.zip](https://github.com/PCSX2/pcsx2/files/7787477/gs_20211229062410.gs.xz.zip)
Savestate: [SLPM-66716 - (F0DCF949).zip](https://github.com/PCSX2/pcsx2/files/7787479/SLPM-66716.-.F0DCF949.zip)

**Issue 2 - Using the PCSX2 Capture feature "Start Video Capture" seems to completely break the textures:**
Video: https://www.youtube.com/watch?v=KUGMcKfe-bM
GS Dump: [gs_20211229062631.gs.xz.zip](https://github.com/PCSX2/pcsx2/files/7787486/gs_20211229062631.gs.xz.zip)
Savestate: [SLPM-66716 - (F0DCF949).zip](https://github.com/PCSX2/pcsx2/files/7787479/SLPM-66716.-.F0DCF949.zip)

Be aware that the games Growlanser 5 and 6 are affected by this bug:
https://github.com/PCSX2/pcsx2/issues/4925
This is also present in the Vulkan Renderer and doesn't affect those 2 issues, so i left it turned off.
Changing "Blending Accuracy" does not seem to affect those issues.
Let me know if you need more information and thank you so much for the Vulkan Renderer!

Tokman5 commented:
Windows 10 Home 21H1
Intel Core i7-8700 @ 3.2GHz
NVIDIA GTX1070 Studio Driver 472.84
PCSX2 version: ``442e1f7`` and ``48f9e0b``
Game: All of ``Jikkyou Powerful Pro Yakyuu`` series I have (``7``, ``8``, ``9``, ``10``, ``11``, ``12``, ``2009``) and ``Jikkyou Powerful Major League 2009``

PCSX2 seems to render only every 2 frames while FMV playing.
this causes screen flickering if deinterlace mode is set to none or bob. If weave tff/bff is set, it is just like a scanline effect. Blend tff/bff make the image darker.

This happens only with Vulkan. OGL, D3D11, and Software don't have this problem.
FMVs of any other games I tried seem to be fine. I think it's because ``Jikkyou ~`` series set different GS register settings from other games.

[gs_20211229152901.gs.zip](https://github.com/PCSX2/pcsx2/files/7787691/gs_20211229152901.gs.zip)
multi frame dump of the opening FMV of ``~ Pro Yakyuu 11`` taken in ``442e1f7``.

dante3732 commented:
I am experiencing a slowdown in fullscreen mode on a vulkan in MGS3. In windowed i have fullspeed.

refractionpcsx2 commented:
> I am experiencing a slowdown in fullscreen mode on a vulkan in MGS3. In windowed i have fullspeed.

What happens in OpenGL?

dante3732 commented:
> What happens in OpenGL?

On vulkan, I have 60 VPS in window mode and 40 in fullscreen mode.
On opengl 24 in the window and 40 in fullscreen mode.
I have amd card.

RIKLAVIEGASINTRODUCTION commented:
MGS2 seems to have an issue with the life bar - it's cut off.

![vk](https://user-images.githubusercontent.com/92535412/147664980-5c2cce52-4725-4f8f-b056-bd44006aaa4b.png)


Vulkan left OGL right

Silanda commented:
Did a quick test of MGS3 and a couple of other games that require higher blending levels to look correct. Performance is generally improved over OpenGL.

I assume you are aware, but just in case: the recently implemented scanmask simulation (#4385) either doesn't appear to be working or hasn't been implemented yet in the Vulkan renderer. 

Also, in Soul Calibur 2 (PAL) this happens on the character select screen with texture preloading enabled:

D:\a\pcsx2\pcsx2\common\Vulkan\Context.cpp(985) : assertion failed:
    Function:  Vulkan::Context::DoSubmitCommandBuffer
    Thread:    MTGS
    Condition: false
    Message:   Failed to submit command buffer.

Stacktrace:
[00] 0x00007FF7CADE1310                          
[01] 0x00007FF7CA6D119A                          
[02] 0x00007FF7CA9F7E37                          
[03] 0x00007FF7CAA2B59A                          
[04] 0x00007FF7CA9D84E3                          
[05] 0x00007FF7CA8025F4                          
[06] 0x00007FF7CADD52E6                          
[07] 0x00007FF7CADD5223                          
[08] 0x00007FF7CADD4DD2                          
[09] 0x00007FF7CADD4DFE                          
[10] 0x00007FF7CAB45C85                          
[11] recalloc                                    
[12] BaseThreadInitThunk                         
[13] RtlUserThreadStart

lightningterror commented on bin/resources/shaders/vulkan/tfx.glsl:
```diff
@@ -0,0 +1,1149 @@
+		#else
+				vec3 D = vec3(0.0f);
+		#endif
+
```
Missing alpha clamp.
https://github.com/PCSX2/pcsx2/blob/master/bin/resources/shaders/opengl/tfx_fs.glsl#L705

dante3732 commented:
When trying to load the game into MGS 3, game give an error, and then the game generally says that the memory cards are not inserted.
![32324](https://user-images.githubusercontent.com/66036249/147674400-4d11dc91-3b49-427d-a1d8-9861c4bede0f.png)

dante3732 commented:
And I still do not have a fullspeed in heavy cutscenes on the vulkan, but on the dx11 fine. Blending use so many resources?
![5456](https://user-images.githubusercontent.com/66036249/147675270-e164c280-329a-44a7-8c70-d6bd1e84a5b5.png)

MrCK1 commented:
> When trying to load the game into MGS 3, game show error, and then the game generally says that the memory cards are not inserted. ![32324](https://user-images.githubusercontent.com/66036249/147674400-4d11dc91-3b49-427d-a1d8-9861c4bede0f.png) ![3](https://user-images.githubusercontent.com/66036249/147674492-5c4a4091-b793-4d15-ac02-75075228c643.png)

There are no changes in this PR that would affect memcards. Please use the forums or Discord for support on that issue.

dante3732 commented:
> There are no changes in this PR that would affect memcards. Please use the forums or Discord for support on that issue.
Apparently some random bug. I saved the game again, and now it loads without problems.

lightningterror commented:
Codacy warnings, as well as some gcc/clang can be taken care of
https://github.com/PCSX2/pcsx2/runs/4659347257?check_suite_focus=true

trostboot commented:
GPU Palette Conversion and Texture preload combined doubles performance in certain scenes in Valkyrie Profile 2, but performance is still significantly behind OGL (palette conversion has no real effect on OGL here) or even the SW renderer. From the opening in-game cinematic:
OGL
![vp2ogl](https://user-images.githubusercontent.com/17594477/147739254-da11b7f6-67d2-4997-a3b7-60e37c077e3d.png)
Vulkan
![vp2vulkan](https://user-images.githubusercontent.com/17594477/147739268-2cf792e6-7d82-48e6-a811-c3fe32a3da2c.png)
Vulkan w/ texture preload and GPU palette conversion
![vp2vulkantexturepreload](https://user-images.githubusercontent.com/17594477/147739274-41b9a0d2-9b2f-4f1b-add9-90fcc3e9621c.png)
Texture preload and palette conversion each add ~20fps in this scene.

GSdump
[gs_20211230102952.zip](https://github.com/PCSX2/pcsx2/files/7792326/gs_20211230102952.zip)

Tested on NV, 1080, driver 497.29, Ryzen 3900X, Win 11 21H2 build 376.0.
Blending options or disabling depth emulation doesn't really affect it.

refractionpcsx2 commented:
I tested your gs dump with default settings but Native x3 with OGL and Vulkan and did not reproduce your results, 

I'm on a 2080Ti, 497.29, Ryzen 3900x, windows 10

My Results:
Native x3
Vulkan: 287 fps
Vulkan + Preload: 338fps
OpenGL: 190fps
OpenGL + Preload: 215fps

trostboot commented:
Performance is definitely higher in the GS debug view. Vulkan palette conversion and preload ~113fps, 45fps without, OGL 300, SW 170 /w 6 threads. Palette conversion or preload has no effect for me in OGL, either. Note that I disabled VRR here, otherwise it was limited to just under 165 to match the refresh rate.

Bit stumped, tried a variety of options, made sure GPU power management is set to max performance for this build, but the results all stay the same for me. GPU usage itself is very low, however.

Trying a few other games, results are fairly similar. Vulkan is consistently much slower for me. Soulcalibur 3 title menu GSdump ([gs_20211230121901.zip](https://github.com/PCSX2/pcsx2/files/7792667/gs_20211230121901.zip)) is ~480 in OGL, ~50 in Vulkan /w preload, only ~17 if palette conversion is also on.
Only thing I can say is that Vulkan performance is fine for me in other applications, so I don't think it's something that's necessarily only on my end, but I'll drop it for now since it seems like an abnormal result.

lightningterror commented:
List of issues I spotted:
Sonic Unleashed crashes on Vulkan
[Sonic Unleashed.zip](https://github.com/PCSX2/pcsx2/files/7793685/Sonic.Unleashed.zip)

Xenosaga dream sequence broken (recommend checking DoMerge feedback write)
[xenosaga_dream_sequence_black.zip](https://github.com/PCSX2/pcsx2/files/7793686/xenosaga_dream_sequence_black.zip)

Fifa Street shadows broken which use DATE
[Fifa Street fbmask.zip](https://github.com/PCSX2/pcsx2/files/7793690/Fifa.Street.fbmask.zip)

Teriminator 3 rock textures broken in the distance
[terminator3r_texture_corruption_nigthvision.zip](https://github.com/PCSX2/pcsx2/files/7793695/terminator3r_texture_corruption_nigthvision.zip)

Beyond good and evil is about 50fps slower on Vulkan compared to OpenGL
[Beyond good and evil shadows.zip](https://github.com/PCSX2/pcsx2/files/7793697/Beyond.good.and.evil.shadows.zip)

Testing was done with preload textures disabled.

stenzek commented:
> Sonic Unleashed crashes on Vulkan

Fixed.

> Xenosaga dream sequence broken (recommend checking DoMerge feedback write)

Fixed.

> Fifa Street shadows broken which use DATE

Fixed.

> Teriminator 3 rock textures broken in the distance

Fixed.

> Beyond good and evil is about 50fps slower on Vulkan compared to OpenGL

When playing back the dump on AMD, I get ~780fps Vulkan, ~730fps DX11, ~360fps GL. But that's AMD GL..

Risae commented:
> **Issue 1 - Flickering of the screen background textures:** Video: https://www.youtube.com/watch?v=bdWb7Rjxexw GS Dump: [gs_20211229062410.gs.xz.zip](https://github.com/PCSX2/pcsx2/files/7787477/gs_20211229062410.gs.xz.zip) Savestate: [SLPM-66716 - (F0DCF949).zip](https://github.com/PCSX2/pcsx2/files/7787479/SLPM-66716.-.F0DCF949.zip)
> 
> **Issue 2 - Using the PCSX2 Capture feature "Start Video Capture" seems to completely break the textures:** Video: https://www.youtube.com/watch?v=KUGMcKfe-bM GS Dump: [gs_20211229062631.gs.xz.zip](https://github.com/PCSX2/pcsx2/files/7787486/gs_20211229062631.gs.xz.zip) Savestate: [SLPM-66716 - (F0DCF949).zip](https://github.com/PCSX2/pcsx2/files/7787479/SLPM-66716.-.F0DCF949.zip)

Using the same settings as before both of those issues seem to be fixed in `PCSX2-64bit-AVX2-pr[5224]-sha[fda9136]-title[[WIP] GS Add a Vulkan renderer]`.
I will see if i can find anything else broken in Vulkan on Growlanser 5 / 6 and write a comment here.

Some other things i noticed:

- The "Preload Textures" Option in the Graphics Settings does not have a tooltip
- The game window resets its position after you open and close the "Graphics Settings": https://www.youtube.com/watch?v=EK_AMp0GcpM
Not sure if this is working as intended. In previous versions of the PCSX2 dev builds the window didn't reset after opening and closing the "Graphics Settings":

HeroponRikiBestest commented:
On the Vulkan renderer, Snoopy vs. the Red Baron experiences an additional HW rendering issue on Vulkan not present in https://github.com/PCSX2/pcsx2/issues/5240. A green light that shouldn't be there is rendered in front of the player.
https://user-images.githubusercontent.com/50224630/147813773-f3c496d5-3b51-4ca3-8ad6-a800f00f29cf.mp4
[dumps.zip](https://github.com/PCSX2/pcsx2/files/7795731/dumps.zip)

HeroponRikiBestest commented:
In monster hunter 2, most text and text boxes are garbled or missing. Some npc text starts out correctly rendered and then becomes garbled after a second or two. This also extends to some ui elements.
GSdumps: [MH2.zip](https://github.com/PCSX2/pcsx2/files/7796042/MH2.zip)

lightningterror commented:
Channel shuffle doesn't work on mission impossible.
[missionimpossible_rocks.zip](https://github.com/PCSX2/pcsx2/files/7796242/missionimpossible_rocks.zip)

stenzek commented:
> Channel shuffle doesn't work on mission impossible.

Fixed.

DonelBueno commented:
The HW hacks aren't implemented yet? Half-pixel offset doesn't work here.

Twisted Metal Black doesn't show the logos at the begning (there is sound, but just a black screen) and the main menu background is also black:

![PCSX2_2021_12_31_15_28_45_871](https://user-images.githubusercontent.com/4239694/147835548-9acddd08-eef8-4644-b1f0-295bd578d809.png)

Geforce GTX 1070, drivers 497.27.

tjpleasant commented:
[gs_20211231133123.gs.zip](https://github.com/PCSX2/pcsx2/files/7796940/gs_20211231133123.gs.zip)
![Capture](https://user-images.githubusercontent.com/81310474/147837212-eb633917-b610-4b18-936d-e4219afcde3b.PNG)

I'm having major graphical issues with Sly Cooper 1, USA region.
Windows 10 Education Version 21H2
Ryzen 5 2400G with a Radeon RX 580
Latest AMD drivers: 21.12.1
This happens with the default settings (aside from changing to Vulkun) with the following build: PCSX2-64bit-AVX2-pr[5224]-sha[ec0c5fe]-title[[WIP] GS Add a Vulkan renderer]

coornio commented:
Would like to give it a try myself, but I can't seem to actually boot into any games whatsoever, regardless of renderer, when testing this PR.

`D:\a\pcsx2\pcsx2\pcsx2\Frontend\ImGuiManager.cpp(58) : assertion failed:
    Function:  ImGuiManager::Initialize
    Thread:    MTGS
    Condition: false
    Message:   Failed to load font data

Stacktrace:
[00] 0x00007FF78B11362F                          
[01] 0x00007FF78B2C384E                          
[02] 0x00007FF78B0EC3C0                          
[03] 0x00007FF78B3C8151                          
[04] 0x00007FF78B0EC570                          
[05] 0x00007FF78B6BFDE6                          
[06] 0x00007FF78B6BFD23                          
[07] 0x00007FF78B6BF8D2                          
[08] 0x00007FF78B6BF8FE                          
[09] 0x00007FF78B430785                          
[10] recalloc                                    
[11] BaseThreadInitThunk                         
[12] RtlUserThreadStart`

This message appears every single time, and disabling the OSD doesn't affect it. It repeats reliably, though I'm unsure about 00 to 09. Anything that could be done about it? Windows 11 by the way. Intel, with nVidia GPU.

5d-0 commented:
With this PR this [issue](https://github.com/PCSX2/pcsx2/issues/257) now can't be fixed even with HW hacks. All api's has this problem in this branch so it's not new renderer problem i assume.

DonelBueno commented:
> With this PR this [issue](https://github.com/PCSX2/pcsx2/issues/257) now can't be fixed even with HW hacks. All api's has this problem in this branch so it's not new renderer problem i assume.

I think HW hacks aren't implement yet, see my post above.

5d-0 commented:
> > With this PR this [issue](https://github.com/PCSX2/pcsx2/issues/257) now can't be fixed even with HW hacks. All api's has this problem in this branch so it's not new renderer problem i assume.
> 
> I thing HW hacks aren't implement yet, see my post above.

They work in other games for me so i dunno.

lightningterror commented:
I took a quick look and it does seem Half pixel offset is broken.

Lolibox commented:
setting interlacing method in GS settings has no effect while a game is running, only the F5 hotkey changes it.

stenzek commented:
HW hack realtime application was broken due to the qt pr this was based on. I've re-done it, so changing in realtime will work fine, as well as keep the GS window open/running.

@coornio you're probably missing files/have an incomplete install. This PR is based on qt-part-2, which adds additional font files.

@tjpleasant Sly 1 is fine on my RX580.

Tokman5 commented:
Is a blockdump more helpful than gsdump? the FMV flickering is also happened on ``ff8cbbb``. Other info is posted in [my previous comment](https://github.com/PCSX2/pcsx2/pull/5224#issuecomment-1002431938).
[blockdump.zip](https://github.com/PCSX2/pcsx2/files/7797734/pawa10blockdump.zip)
after loading memcards, just wait for FMV playing.

refractionpcsx2 commented:
> Is a blockdump more helpful than gsdump? the FMV flickering is also happened on `ff8cbbb`. Other info is posted in [my previous comment](https://github.com/PCSX2/pcsx2/pull/5224#issuecomment-1002431938). [blockdump.zip](https://github.com/PCSX2/pcsx2/files/7797734/pawa10blockdump.zip) after loading memcards, just wait for FMV playing.

I don't think you meant to post on this pr

Tokman5 commented:
> I don't think you meant to post on this pr

I think I'm talking about a Vulkan issue. Is something wrong?

refractionpcsx2 commented:
Sorry, was looking at stuff on my phone and there was an FMV flickering issue just made, didn't realise it was somebody else with a different issue.

HeroponRikiBestest commented:
Happy! Happy!! Boarders in Hokkaidou Rusutsu Resort appears to suffer from the same garbled text issue as Monster Hunter 2, although it affects very few text boxes in comparison.
[gs_20220101071551.gs.zip](https://github.com/PCSX2/pcsx2/files/7797814/gs_20220101071551.gs.zip)

nobleflame commented:
I've been playing around with MGS2 and I have to say that this is a massive improvement over OpenGL in terms of performance. 

@tellowkrinkle suggested I try this as I was getting slow down during specific speeds with blending set to high on OpenGL. Using the Vulkan renderer with blending set to high keeps a locked 60fps in all of the scenes that I've tested so far.

Silanda commented:
The animated loading icon in James Bond 007: Agent Under Fire (PAL) appears as a solid colour square rather than the animated icon that it should be (and is in every other renderer).
[AgentUnderFireVulkanBug.zip](https://github.com/PCSX2/pcsx2/files/7798298/AgentUnderFireVulkanBug.zip)

stenzek commented:
@HeroponRikiBestest looks fine to me on AMD... maybe a driver bug?

![image](https://user-images.githubusercontent.com/11288319/147865060-1627afae-50f1-4f3b-a3e2-d3db422d1604.png)
![image](https://user-images.githubusercontent.com/11288319/147865064-5bd3b4de-5515-4ff3-872f-373979701c59.png)

HeroponRikiBestest commented:
@stenzek I'm also using AMD, albiet on Linux with mesa.

HeroponRikiBestest commented:
Okay, I can confirm the issue was caused by an out of date MESA version. All of my previously posted issues are invalid. I'm really sorry about that, I thought I was up to date.

stenzek commented:
All good, at least we know what the solution is when someone complains.

stenzek commented:
> The animated loading icon in James Bond 007: Agent Under Fire (PAL) appears as a solid colour square rather than the animated icon that it should be (and is in every other renderer).

Fixed.

DonelBueno commented:
Twisted Metal Black is fully fixed now, great job!

The frame pacing on the Vulkan renderer is great, much butter than on OpenGL, even GT4 is perfectly fluid.

gilderoylockhart01 commented:
To be short, on ampere cards this Vulkan render saves the day. It is much much faster than GL when using blending. I could actually play ZOE 2 at full speed with 8x and full blending to give you an idea, and the card does this fingers in the noise.

There is also a few performance notes which I noticed:

- I can see a serious improvement when it comes to vram spikes, I see a lot less of "out of vram" logs when running in vulkan with high resolution scaling.

- Auto flush speed impact is almost non-existant with vulkan (except the games that are in the meta for slow auto-flush games) which is a good thing because it's more correct to have it on by default, was that intended?

- Games that does less draw calls have seen a quite impressive performance bump, but that "performance per draw calls" scaling is not linear, if the game does a "bunch" of draw calls (example, ONI which runs at a completely unlocked framerate and can reach silly FPS on real hardware when on the main menu) is not improved by a lot when using Vulkan.

- Full blending is in some cases twice as fast on ampere (example MGS 3) I wonder if turing is the same.

There is just 1 game which is an exception to the rule: "Black". When using full blending the framerate absolutely tanks like it does on GL. GSdump:
[Black Vulkan Blending.zip](https://github.com/PCSX2/pcsx2/files/7799061/Black.Vulkan.Blending.zip)

gilderoylockhart01 commented:
Ooops, it turns out there is a problem with "Monkey Ball deluxe". The loading screen have issues in the vulkan render:

OpenGL:
![Capture d’écran 2022-01-02 135646](https://user-images.githubusercontent.com/90222976/147876476-a4973aa5-4c02-4ca4-bd6d-2ca5533cfa6e.png)

Vulkan:
![Capture d’écran 2022-01-02 135630](https://user-images.githubusercontent.com/90222976/147876479-84b1e349-4ca0-4acc-8b55-59cb130c63f4.png)

GSdump:
[Monkey Ball Vulkan issue.zip](https://github.com/PCSX2/pcsx2/files/7799088/Monkey.Ball.Vulkan.issue.zip)

stenzek commented:
> The frame pacing on the Vulkan renderer is great, much butter than on OpenGL, even GT4 is perfectly fluid.

That makes zero sense. It should be exactly the same, the renderer doesn't control when presents happen.

refractionpcsx2 commented:
@gilderoylockhart01 if you haven't can you please make sure your drivers are up to date since we've had s few bugs caused by old nvidia drivers

gilderoylockhart01 commented:
I am on the 497.29 driver version, it doesn't ask for any update so I guess it's updated.

refractionpcsx2 commented:
Yes that's up to date, just wanted to be sure so we weren't chasing ghosts

stenzek commented:
Monkey ball fixed with the latest push, as well as Tales of Abyss causing shader compile errors (pointed out by bositman).

TheLastRar commented:
The screenshot/snaps hotkey (F8) doesn't save a screenshot, logging the following
```
Snapshot taken
GSTextureVK::Save not implemented
```

Lolibox commented:
Using the pause hotkey while the GS window is fullscreen can cause it to hang/freeze instead of closing instantly like it does while it is windowed.

Tokman5 commented:
``a216773`` doesn't fix [FMV of Power Pros series flickering](https://github.com/PCSX2/pcsx2/pull/5224#issuecomment-1002431938).

refractionpcsx2 commented:
> `a216773` doesn't fix [FMV of Power Pros series flickering](https://github.com/PCSX2/pcsx2/pull/5224#issuecomment-1002431938).


I can't see the flickering in the gsdump because the dump runs at 500fps lol, but it's certainly darker than OGL/DX

Tokman5 commented:
> I can't see the flickering in the gsdump because the dump runs at 500fps lol over 2x OGL)

Can you try a blockdump?
[blockdump.zip](https://github.com/PCSX2/pcsx2/files/7797734/pawa10blockdump.zip)

refractionpcsx2 commented:
I can't do anything with that blockdump, it runs out of block as soon as it hits the intro fmv, if I skip that it runs out of blocks on the menu.

Tokman5 commented:
hmm, i'll try to recreate a blockdump later.

Tokman5 commented:
It seems to be fixed in ``c057e8f``. FMVs are rendered fine.
the edges of the FMV screen are still flickering, but that is also happened in other HW renderers (OGL, D3D11), so I think that is the different issue.

lightningterror reviewed (approved):
Now needs extensive testing, otherwise ready.

JordanTheToaster reviewed (approved):
I've tested 70 games and found no issues. I have lost control of my life.

RedDevilus commented:
So while trying to play rogue galaxy, I had a small moment of IPU(FMV) corruption. Turns out that because I have a laptop with outdated Intel Driver it picked DX11 SW then I updated and got OpenGL HW and it does use my dedicated GPU but the corruption is still there, so have to start in the individual renderers to avoid it.

So forcing NVIDIA GPU it works fine again:
![image](https://user-images.githubusercontent.com/24227051/148091775-c5c61871-5608-4675-9b1f-690cafe54eef.png)
![image](https://user-images.githubusercontent.com/24227051/148091798-6da5c2c9-4e8a-410d-933a-dd78506bd05c.png)

When NVIDIA auto-selects to Integrated:
![image](https://user-images.githubusercontent.com/24227051/148091860-8cf8805a-e70a-409b-b189-903818d1f100.png)
![image](https://user-images.githubusercontent.com/24227051/148092027-4ad0b6cf-75c1-41af-83ae-1206ec4e26e0.png)
https://user-images.githubusercontent.com/24227051/148092076-76fb4b7c-fa15-4106-96a2-e735fbeaf7af.mp4

So basically it's getting confused between 2 adapters because of not populating it correctly.

jeremybris commented:
I tested about 20 games on the latest branch, and here are the only regressions I found:

1. Superman Returns sky color is off and the game flickers. (It went in-game for the first time recently)
2. Call of Duty 2: Big Red One and Spider-Man 2, both with broken mipmapping and using a similar in-game engine, crash (if I am testing this correctly. I downloaded PCSX2-32bit-AVX2). 

I have an RTX 2080 Super and Ryzen 9 3900x. Thank you for your work.

![Screenshot 2022-01-03 210132](https://user-images.githubusercontent.com/11889359/148000391-ce6c5d13-9576-466f-a579-e05eef1fe8ca.jpg)

![Screenshot 2022-01-03 210823](https://user-images.githubusercontent.com/11889359/148000424-d9732727-316b-44ce-b11d-5d3b66d59e63.jpg)

![Screenshot 2022-01-03 203409](https://user-images.githubusercontent.com/11889359/147998183-0d99128a-3952-41fc-8860-eeb95cbd2803.jpg)

lightningterror commented:
> I tested about 20 games, and here are the only regressions I found:
> 
> 1. Superman Returns sky color is off and the game flickers. (It recently became more playable.)
> 2. Call of Duty 2: Big Red One and Spider-Man 2, both with broken mipmapping and using a similar in-game engine, crash (if I am testing this correctly. I downloaded PCSX2-32bit-AVX2).
> 
> I have an RTX 2080 Super and Ryzen 9 3900x. Thank you for your work.
> 
> ![Screenshot 2022-01-03 210132](https://user-images.githubusercontent.com/11889359/148000391-ce6c5d13-9576-466f-a579-e05eef1fe8ca.jpg)
> 
> ![Screenshot 2022-01-03 210823](https://user-images.githubusercontent.com/11889359/148000424-d9732727-316b-44ce-b11d-5d3b66d59e63.jpg)
> 
> ![Screenshot 2022-01-03 203409](https://user-images.githubusercontent.com/11889359/147998183-0d99128a-3952-41fc-8860-eeb95cbd2803.jpg)

GS dumps would be nice so Sten can look at the issues.

jeremybris commented:
> GS dumps would be nice so Sten can look at the issues.

I would like to provide a GS dump, but I don't know how. Sorry I am new.

RedDevilus commented:
Shift + F8 to make a GS Dump, it will be in your snaps folder. Then just zip and upload here.

jeremybris commented:
> Shift + F8 to make a GS Dump, it will be in your snaps folder. Then just zip and upload here.

Thank you! I did not know how. Superman Returns: https://drive.google.com/file/d/1rRsQGJD6EevT9Fcld2yw-g6E5JjYCmOc/view?usp=sharing .

Spider-Man 2 and Call of Duty 2 Big Red One just crash, so I can not provide a GS Dump.

lightningterror commented:
> Spider-Man 2 and Call of Duty 2 Big Red One crash, so I don't think I can provide a GS Dump.

You can take the dump using a different renderer like opengl or dx11, will still do it's job.

jeremybris commented:
> > Spider-Man 2 and Call of Duty 2 Big Red One crash, so I don't think I can provide a GS Dump.
> 
> You can take the dump using a different renderer like opengl or dx11, will still do it's job.

Spider-man 2 OpenGL: https://drive.google.com/file/d/1pCy8JEjhD3Xa7GC1i671cFSpcthV1DpM/view?usp=sharing

Call of Duty 2 OpenGL: https://drive.google.com/file/d/1LN5g0W5YHxi5676AN6lmhThDoZe2DFS4/view?usp=sharing

![gs_20220104163025](https://user-images.githubusercontent.com/11889359/148128614-2790068b-3e3a-4120-9d38-04e2c3ee7adb.png)

![gs_20220104163351](https://user-images.githubusercontent.com/11889359/148128692-156d6a93-6d8d-4db7-9993-584878043955.png)

stenzek commented:
COD2 was doing an out of bounds copy (also affected DX11).

Superman should be fixed now.

Silanda commented:
Is anyone else having a problem with the manual "Switch to Software renderer for FMVs" game fix not working with this?

jeremybris commented:
> COD2 was doing an out of bounds copy (also affected DX11).
> 
> Superman should be fixed now.

Thank you! Also, maybe not worth fixing? But the "screenfix" hack I got [here](https://www.youtube.com/watch?v=52VzL_5ariY) for Mercenaries crashes in the main menu in Vulkan, whereas it does not in OpenGL.

Cheat/Hack: 
[23510F99.zip](https://github.com/PCSX2/pcsx2/files/7815640/23510F99.zip)

[OpenGL GS Dump](https://drive.google.com/file/d/10D8kf83qqiXEXIAypHU3qCJpP9DRbDOR/view?usp=sharing)

coornio commented:
Not sure if anyone else faced this, but when testing this PR out, twice now and with no idea why or how, the main game window would stop responding to ALL keyboard input, be it hotkeys or controller input. Even escape wouldn't work to close it, I had to go through the main app window. I am not 100% sure if this is related to vulkan or some side effect of the Qt merge branch part 2 a short while back.

That said, so far I have not come across any particular issues in any games I've tested, though I haven't been through my entire library either yet. Nothing but praises to sing for now!

gtgamer468 commented:
Tested Gran Turismo 3, 4 Prologue, and 4 alongside Tekken 5 and Midnight Club 3. I noticed a bit of a speed improvement especially on night tracks (even with high blending on).

Daisouji commented:
Test Airblade, Vulkan works well on 1050 pretty well compared OGL. On my Intel 630, It crashes with "Failed to submit command buffer"  for this game only.

lightningterror reviewed (commented):
A couple of new warnings, below a bit too
https://github.com/PCSX2/pcsx2/runs/4738398122?check_suite_focus=true#step:11:644

and some codacy stuff
https://github.com/PCSX2/pcsx2/pull/5224/checks?check_run_id=4738418372

And we are good to go.

DonelBueno commented:
Some hotkeys aren't working here:

PAGE_UP    - FXAA antialiasing (HW and SW)
PAGE_DOWN  - Cycles through dither modes (HW) 
DELETE     - Software Antialiasing (AA1)
HOME       - FX shader
INSERT     - Hardware mipmapping

The only one that works is F5 (Deinterlace modes).

refractionpcsx2 commented:
They're gone, there's no real reason you should need those on the fly.

RedDevilus commented:
In the future it would just be easier to have a list of keybindings that you want to assign instead of the odd hardcoded way.

Silanda commented:
Are the FXAA and external shader working at all with Vulkan currently? I see a difference under OpenGL but not Vulkan.

DonelBueno commented:
> Are the FXAA and external shader working at all with Vulkan currently? I see a difference under OpenGL but not Vulkan.

From my experience, no.

RedDevilus commented:
Fixes https://github.com/PCSX2/pcsx2/issues/1047

nobleflame commented:
This is great, thanks to all involved for this. It's massively improved performance.

Ghaleon7771 commented:
I dont know if i downloaded a bad build... i am able to select Vulkan as a driver, but i see no major performance VS Open GL. I am using a R7 360 video card.

refractionpcsx2 commented:
it depends if you're limited by the GS/GPU.

but OpenGL should be slower than Vulkan, if that's what you're limited on. But going off your GPU, I'm not expecting your CPU to be too great.

Ghaleon7771 commented:
> it depends if you're limited by the GS/GPU.
> 
> but OpenGL should be slower than Vulkan, if that's what you're limited on. But going off your GPU, I'm not expecting your CPU to be too great.

Yeah, my PC is s**** its old and outdated, the CPU is a I3 3240, i cant expect miracles with this garbage PC, so you are right. The issue is my poor´s CPU/GPU

OCRBonk commented:
Not sure if relevant since it requires a hack but 
**Xenosaga I** requires `Half Pixel Offset: Special (Texture)` in order to fix the shadow puppet lines *and display shadows in cutscenes properly. Otherwise you get strange clipping glitches. 
This only works in OpenGL but the blending can require much more performance out of the GPU.
At 4x res on an RTX 2080 just in this one scene GPU usage can drop by 15-20% on Vulkan vs OpenGL. (And that just at the Basic Blending level.  Can be as high as a 30% drop on Ultra)
But with Vulkan you can't seem to get rid of the shadow puppet lines *and* get correct shadows in Cutscenes. Same as in DX11. Only 1 or the other.
Software:
![pcsx2-avx2_2022_01_08_22_11_05_545](https://user-images.githubusercontent.com/16983740/148671647-6d02970d-e509-41d2-91a6-fed5fa00e83a.png)
OpenGl with HPO:S(T)
![pcsx2-avx2_2022_01_08_22_09_52_666](https://user-images.githubusercontent.com/16983740/148671655-3cc53e3a-9c6b-4b58-9f21-6ea807f16235.png)
DX11
![pcsx2-avx2_2022_01_08_22_11_48_101](https://user-images.githubusercontent.com/16983740/148671662-9a5c0f0f-59db-4410-8031-c4af2a57098a.png)
Vulkan
![pcsx2-avx2_2022_01_08_22_10_34_191](https://user-images.githubusercontent.com/16983740/148671669-c6a7487e-3db5-44e0-87a7-14a99a1d12bc.png)

brad86m commented:
Holy shit at the input latency reduction with Vulkan. This is the most important advancement to me. The emulator has been plagued for year and years with such high latency that made so many games near unplayable. Fighting, or action games like DMC and God Hand felt such a chore to play. The latency pretty much stripped all the fun from otherwise superb games, and didn't make the emulation route viable for anything that wasn't slower paced, like horror and RPG's. Orignal members of the team didn't seem interested in implementing exclussive fullscreen, or trying to remedy the latency issues, after so many years of users begging for it. Recent changes to the team, and some skilled fresh blood willing to help with the emualtor has started to make a big difference.  
The recent MTGS option was a start, but that only helped to a small degree. Vulkan on the other hand seems to have made an absolute massive difference.

Really happy with this. Thank you.

stenzek commented:
The renderer has nothing to do with input latency.

brad86m commented:
Apologies. I didn't mean to aim that at the majority of team members. Going by what I have witnessed from mainly many years ago. On the old forum before Discord, and other modern ways of communicating on these topic. Some older team members were really arrogant to people that asked them even basic things. Sure, they owed us absolutely nothing, and their time put into the project was always appreciated, but I never understood why there was so much shitty communication between those devs and users.
That's history, though. Things are looking up, big time.

ForeskinJohn commented:
will vulkan be in later devbuilds? if so this will make games like gran turismo 4 and shadow of the colossus run amazingly

F0bes commented:
It is in dev builds, the PR has been merged.
